substrafl/algorithms/pytorch/torch_fed_pca_algo.py

```diff
@@ -0,0 +1,277 @@
+import math
+from pathlib import Path
+from typing import Any
+from typing import List
+from typing import Optional
+
+import torch
+
+from substrafl import NpIndexGenerator
+from substrafl.algorithms.pytorch import weight_manager
+from substrafl.algorithms.pytorch.torch_base_algo import TorchAlgo
+from substrafl.remote import remote_data
+from substrafl.schemas import FedAvgAveragedState
+from substrafl.schemas import FedAvgSharedState
+from substrafl.schemas import StrategyName
+
+
+class TorchLinearModel(torch.nn.Module):
+    """Define linear model to encapsulate eigenvectors.
+
+    Args:
+        in_features (int): dimension of input vectors
+        out_features (int): number of dimensions to keep as part of dimensionality
+            reduction
+    """
+
+    def __init__(self, in_features: int, out_features: int):
+        super().__init__()
+        self.in_features = in_features
+        self.out_features = out_features
+        self.eigen_vectors = torch.nn.Linear(self.in_features, self.out_features, bias=False)
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        """Perform dimensionality reduction.
+
+        Args:
+            x (torch.Tensor): inputs to map to reduced dim space.
+                shape of x is (B, D_IN)
+
+        Returns:
+            torch.Tensor: reduced dim vectors
+        """
+        eigen_vectors = self.eigen_vectors.weight.data
+        coefs = x @ eigen_vectors.T
+        return coefs @ eigen_vectors
+
+
+class TorchFedPCAAlgo(TorchAlgo):
+    """TorchFedPCAAlgo class, inheriting from TorchAlgo and designed to perform Principal Component Analysis (PCA).
+
+    Args:
+        dataset (torch.utils.data.Dataset): input data on which to perform PCA
+        in_features (int): input data dimensionality
+        out_features (int): number of dimensions to keep after PCA
+        batch_size (Optional[int]): mini-batch size
+        seed (int): random generator seed. Default to 1.
+        use_gpu (bool): whether to use GPU or not
+    """
+
+    def __init__(
+        self,
+        dataset: torch.utils.data.Dataset,
+        in_features: int,
+        out_features: int,
+        batch_size: Optional[int] = None,
+        seed: Optional[int] = None,
+        use_gpu: bool = True,
+        *args,
+        **kwargs,
+    ):
```

Comment:
```suggestion
    def __init__(
        self,
        *,
        dataset: torch.utils.data.Dataset,
        in_features: int,
        out_features: int,
        batch_size: Optional[int] = None,
        seed: Optional[int] = None,
        use_gpu: bool = True,
        **kwargs,
    ):
```
I'm gonna start a crusade to enforce kwargs on new functions with more than 3 parameters 🙃